lib/spack/spack/binary_distribution.py

```diff
@@ -935,15 +935,22 @@ def build_tarball(spec, outdir, force=False, rel=False, unsigned=False,
     spackfile_path = os.path.join(
         cache_prefix, tarball_path_name(spec, '.spack'))
 
-    remote_spackfile_path = url_util.join(
+    outdir = mirror.push_url
+    fetch_spackfile_path = url_util.join(
+        mirror.fetch_url, os.path.relpath(spackfile_path, tmpdir))
+    push_spackfile_path = url_util.join(
         outdir, os.path.relpath(spackfile_path, tmpdir))
 
     mkdirp(tarfile_dir)
-    if web_util.url_exists(remote_spackfile_path):
+    if web_util.url_exists(push_spackfile_path):
```

Comment:
Can't it be done like:

```
if web_util.url_exists(fetch_spackfile_path):
    web_util.remove_url(push_spackfile_path)
```

I left out the force bit force bit for brevity.  This is what I imagined in my previous review: Since those should be the same file, read from the fetch url, write to the push url.

Reply:
From what you've said, push/fetch feature was designed around the idea that while the URL's are different, they are effectively the same place. I think they don't necessarily have to be, and am in fact counting on that. By letting them be two distinct unconnected spaces, it allows you to set up a mirror such that buildcache will only create binary cached builds you don't already have at a given mirror, and storing them wherever you want them to go. 

In our case, we then push those builds to that same mirror server, but that doesn't necessarily have to be the next step.

This is a bit of a concept change with this feature, I guess. It means you can't assume that a file at one url exists at the other, or that deleting from the push location deletes the file at the fetch location. This CAN be true, but doesn't have to be.

Reply:
There's a design assumption that the two urls present a consistent view of the same build cache.  You are correct that we don't currently do anything in the spack code that strictly requires that they do, but that could change.

At a minimum, if spack pushes a binary to the build cache via the push url, it's expected that the same binary would then be found at the fetch url, and if something is deleted via the push url, it would no longer be found under the fetch url.

Reply:
I should further clarify: what I wrote above is only the intended semantics.  You can certainly point the two urls at completely different locations, but users should do so knowing that spack may treat those locations as consistent.  If they're not consistent, this will produce unintended side effects, and while in this example use case, you actually wanted those unintended effects, we can't support that case as it would likely make future improvements that would strictly require consistency impossible.

Your use case is better modeled by two different build caches, rather than a single cache with different urls.  And your desired behavior sounds like some kind of new feature where Spack would use one cache as a mirror or caching layer for another.  I would enthusiastically support something like this, but not with the current implementation, as it relies on an unintended feature of push and fetch urls, and I don't think we should be formally supporting that feature.